### DISCUSSION TITLE
what is LVal?

### DISCUSSION THREAD

**Discussion by k644606347** [babel/babel#15622] [Q&A] [answered]
The id under VariableDeclarator is of type LVal;
<img width="505" alt="image" src="https://github.com/babel/babel/assets/6951527/848cb88b-a001-42cb-a5f9-fd564044ec56">
```type LVal = Identifier | MemberExpression | RestElement | AssignmentPattern | ArrayPattern | ObjectPattern | TSParameterProperty | TSAsExpression | TSSatisfiesExpression | TSTypeAssertion | TSNonNullExpression;```

I only know one case: ```VariableDeclarator. id===Identifier```:
<img width="1615" alt="image" src="https://github.com/babel/babel/assets/6951527/a5b6c1d5-c73e-4019-a122-6624e640dfc6">

What other cases are there? thanks

**Comment by k644606347** [comment]
Do you have any relevant documents? I may have missed it

**Reply by JLHwung** [reply, marked as answer]
https://babel.dev/docs/babel-types#lval

`LVal` is everything that can fit in the `%` in `var % = init` and `for ( % of arr)`, such as

```ts
var [ x ] = init
var { x } = init
var { ...x } = init
for (x.y of arr) {}
```
